### ISSUE TITLE
Some thoughts about storage improvements.

### ISSUE THREAD

**Issue by cube2222** [conprof/conprof#26]
Hey, it’s me again :)

I’ve been thinking about my storage concerns recently (#11 ) and come up with a few ideas. (So far I’ve been successfuly using my naive file based solution with a few locking optimizations, running it with a one week retention / 1 TB of data without tracing on).

Firstly, some stuff_assumptions_opinions which you maybe can confirm/deny, or I’ll have to benchmark to check:

1. We know the current implementation won’t work with a medium/big cluster (200 instances getting scraped for 10s every 1-2 mins) because it’ll quickly explode in RAM, loading all profiles into memory.
2. One instance generates 10 series. This is - at least in my experience - a lot less than the usual metric series count per instance.
3. One datapoint is comparatively **much** bigger, than with metrics. A cpu/mem profile is tens of kilobytes, a trace is tens of megabytes - as opposed to a float being 64 bytes.
4. @brancz mentioned inode exhaustion issues with a file-per-series approach in standard metric-scraping prometheus.
	1. Do you possibly have a link to a discussion/benchmarks regarding this?
	2. Taking into account points 1-3, won’t other hurdles come in before inode exhaustion with traces being scraped?
	3. Taking into account points 1-3, won’t other hurdles come in before inode exhaustion with traces **not** being scraped? This should also take into account, that one week of raw data with 200 instances 10s per minute is ~1TB
5. Let’s say we want to query long timespans. At query time, we are only interested in metadata. So the datapoint sizes will negatively impact the hot-path performance, even though we don’t need them here (as opposed to metric-scraping prometheus, where we are aggregating data over time while querying).
6. Only when viewing pprof's do we need the actual datapoints, and only one / a few at a time.
7. What I think is definitely a pity with my previous file-based implementation, is that label matching and similar stuff must be done explicitly, instead of being handled by the underlying prometheus dependencies.

So overall I propose two possible approaches:
1. Create one file per datapoint. Store the file paths in TSDB (as opposed to my previous approach). So that TSDB only handles small strings.
2. Assuming we are doing only one scrape per series in parallel. Create an appendable file for each series and store path+offset in TSDB. Path could be stored as a label.

With approach #1 we could use something like [gocloud/blob](https://godoc.org/gocloud.dev/blob) to provide a few file storage backends (file-based, s3, etc.) basically for free.

Both approaches also come without any locking.

What do you think about this? If you have any initial cautions, answers, or think it’s an outright bad direction, please tell me! (Don’t be afraid to be harsh)

**Comment by brancz** [comment]
(I'm about to go on vacation, so I apologize in advance for delayed responses after this one)

First of all, I think it's super awesome that you are getting involved and want to improve things, I very highly appreciate the effort you are putting in and I want to work with you to improve this situation! :)

The way I've been thinking about it is: Prometheus TSDB has some of the exact the same problems, just on a much smaller scale in regards to the sample value size. Prometheus TSDB WAL already writes segments of 128Mb on disk, but keeps the data in-memory. My idea around this was if we could mmap the segments instead of just using them to re-create the in-memory representation of the WAL then we would only have things in memory that we are actually querying + at most 128Mb. To be clear, I realize that "just mmap-ing" is unlikely to be enough, it might need us to rethink the segment format a bit to allow for a strategy like this as well as adapt the read path to potentially be aware of this, kind of as a way of layered caches.

@bwplotka @krasi-georgiev you know the TSDB and WAL code better than I do, do you think this could be feasible?

**Comment by cube2222** [comment]
Just saying, I'm also concerned about the read path. Because there's a lot of "bytes" put into the data structure we use for querying, even though it doesn't need to be there. (later we do a single get for the single datapoint of interest)

(reiterating this as I'd like to hear their thoughts on this too, because maybe I'm trying to fix a problem that just isn't there)

**Comment by krasi-georgiev** [comment]
could you save me a bit of time digging through the conprof code  and give an example of the data format that you are saving in tsdb at the moment?

Also what are the most common queries to get that data.

**Comment by cube2222** [comment]
I think that's the code https://github.com/conprof/tsdb/commit/70f0d4af7ef43d4b2c1f333d3cb7a1a93b0c5467

We're saving go pprof/trace files, which are from the tsdb point of view, slices of bytes containing from a few tens of Kilobytes, up to a few tens of Megabytes.

I'm not sure what you mean with most common queries:
usually you want to scan some small timespan for datapoints with the given labels, but are not interested in the actual data stored there. Then you select the datapoint to open the pprof. That's when you get an iterator to the wanted series and seek to the single timestamp you're interested in and read the bytes there.
```
storageFetcher := func(_ string, _, _ time.Duration) (*profile.Profile, string, error) {
		var prof *profile.Profile

		q, err := p.db.Querier(0, math.MaxInt64)
		if err != nil {
			level.Error(p.logger).Log("err", err)
		}

		ss, err := q.Select(m...)
		if err != nil {
			level.Error(p.logger).Log("err", err)
		}

		ss.Next()
		s := ss.At()
		i := s.Iterator()
		t, err := stringToInt(timestamp)
		if err != nil {
			return nil, "", err
		}
		i.Seek(t)
		_, buf := i.At()
		prof, err = profile.Parse(bytes.NewReader(buf))
		return prof, "", err
	}
```

Currently there are no aggregations.

**Comment by krasi-georgiev** [comment]
yeah I think the simplest solution would be to just keep a reference to the file path in tsdb and when needed just open the file from disk rather than keep the raw bytes in memory.

**Comment by brancz** [comment]
I think that's actually pretty reasonable, essentially using tsdb just as an index. In order for that to work though we absolutely need isolation/MVCC in tsdb otherwise we're gonna have a bad time. I know it's been on our plate for a long time, maybe it's time to finally finish it.

**Comment by krasi-georgiev** [comment]
Why would isolation/MVCC be required? the access to the files would be read only so more than a single process can read from the same file.

**Comment by brancz** [comment]
No when inserting we need isolation between it showing up in the index and appending to the series.

**Comment by cube2222** [comment]
Could you please expand on the problem? As I see it you'd first write the file to some storage layer (file system, object storage, etc.) and after it's closed (and immutable) write its file path to TSDB, the same way you write a byte slice now.

**Comment by brancz** [comment]
Yes you're right, doing it in that order should solve that problem.
